gobblin-core/src/main/java/gobblin/source/DatePartitionedSource.java

```diff
@@ -0,0 +1,279 @@
+import gobblin.source.workunit.MultiWorkUnitWeightedQueue;
+import gobblin.source.workunit.WorkUnit;
+
+
+/**
+ * Implementation of {@link gobblin.source.Source} that reads over date-partitioned daily data. The watermark for each
+ * run is the day that the job read data up to. It uses {@link WorkUnit#getHighWaterMark()} to get the watermark from
+ * the previous workunits. The job has the option of wrapping around once it reaches a certain date. Currently, each
+ * file corresponds to one work unit.
+ */
+public class DatePartitionedSource extends FileBasedSource<Schema, GenericRecord> {
+
+  // Configuration parameters
+  private static final String DATE_PARTITIONED_SOURCE_PREFIX = "date.partitioned.source.";
+  private static final String DATE_PARTITIONED_SOURCE_MIN_WATERMARK_VALUE = DATE_PARTITIONED_SOURCE_PREFIX
+      + "min.watermark.value";
+  private static final String DATE_PARTITIONED_SOURCE_MAX_FILES_PER_JOB = DATE_PARTITIONED_SOURCE_PREFIX
+      + "max.files.per.job";
+  private static final String DATE_PARTITIONED_SOURCE_MAX_WORKUNITS_PER_JOB = DATE_PARTITIONED_SOURCE_PREFIX
+      + "max.workunits.per.job";
+
+  private static final int DEFAULT_DATE_PARTITIONED_SOURCE_MAX_FILES_PER_JOB = 2000;
+  private static final int DEFAULT_DATE_PARTITIONED_SOURCE_MAX_WORKUNITS_PER_JOB = 500;
+  private static final String DEFAULT_DATE_PARTITIONED_SOURCE_DAILY_FOLDER_NAME = "daily";
+
+  // Joda formatters
+  private static final DateTimeFormatter WATER_MARK_FORMATTER = DateTimeFormat.forPattern("yyyyMMdd");
+  private static final DateTimeFormatter DAILY_FOLDER_FORMATTER = DateTimeFormat.forPattern("yyyy/MM/dd");
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatePartitionedSource.class);
+
+  private static final String AVRO_SUFFIX = ".avro";
+
+  private FileSystem fs;
+
+  @Override
+  public Extractor<Schema, GenericRecord> getExtractor(WorkUnitState state) throws IOException {
+    return new DatePartitionedExtractor(state);
+  }
+
+  @Override
+  public void initFileSystemHelper(State state) throws FileBasedHelperException {
+    this.fsHelper = new HadoopFsHelper(state);
+    this.fsHelper.connect();
+  }
+
+  private void init(SourceState state) {
+    DateTimeZone.setDefault(DateTimeZone.forID("America/Los_Angeles"));
+    try {
+      initFileSystemHelper(state);
+    } catch (FileBasedHelperException e) {
+      Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+    init(state);
+
+    // Initialize filesystem variables
+    HadoopFsHelper fsHelper = (HadoopFsHelper) this.fsHelper;
+    this.fs = fsHelper.getFileSystem();
+
+    // Get the starting date
+    LOG.info("Retrieving the low water mark based on the previous execution");
+
+    DateTime lowWaterMark = getLowWaterMark(state);
+
+    LOG.info("The low water mark for this run is " + lowWaterMark);
+
+    // Determine how many files to process
+    int maxFilesPerJob =
```

Comment:
Please move the reading of all configuration properties to the initialization function.